File: binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def zigzagLevelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        answer = []
        if not root:
            return answer

        temp, queue, flag = [], [root], 1
        while queue:
            for _ in range(len(queue)):
                node = queue.pop(0)
                temp.append(node.val)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            answer += [temp[::flag]]
            temp = []
            flag *= -1
        return answer

File: binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution 
{
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) 
    {
        List<List<Integer>> answer = new ArrayList<>();
        if (root == null) return answer;
        
        // Setup for BFS
        Deque<TreeNode> queue = new LinkedList<>();
        queue.addLast(root);
        boolean flag = false;
        
        // BFS
        while (!queue.isEmpty())
        {
            int levelSize = queue.size();
            List<Integer> current = new ArrayList<>();
            
            for (int i = 0; i < levelSize; i++)
            {
                TreeNode node = queue.removeFirst();
                
                if (!flag)
                    current.add(node.val);
                else
                    current.add(0,node.val);
                
                if (node.left != null) queue.addLast(node.left);
                if (node.right != null) queue.addLast(node.right);
            }
            
            flag = !flag;
            answer.add(current);
        }
        
        return answer;
    }
}